cpp/src/join/joining.cu

```diff
@@ -253,260 +220,345 @@ gdf_error join_call( int num_cols, gdf_column **leftcol, gdf_column **rightcol,
+ * from tables `left` and `right` respectively, which are supposed to be joined onto,
+ * that are "in common". For "common" columns, only a single output column will
+ * be produced. Depending on type of join, the values will be gathered.
+ * @param[in] left_indices Row indices from `left` to gather. If any row index
+ * is out of bounds, the contribution in the output `table` will be NULL.
+ * @param[in] right_indicess  Row indices from `right` to gather. If any row
+ * index is out of bounds, the contribution in the output `table` will be NULL.
+ * 
+ * @returns Result of joining `left` and `right` tables as per `left_indices` and
+ * `right_indices`. For any "common" columns indicated by `columns_in_common`,
+ * only a single output column is produced. The resulting table will be joined
+ * columns of `left` and then joined columns of `right` in the same order.
+ */
+/* ----------------------------------------------------------------------------*/
 
 template <JoinType join_type, typename index_type>
-gdf_error construct_join_output_df(
-        std::vector<gdf_column*>& ljoincol,
-        std::vector<gdf_column*>& rjoincol,
-        gdf_column **left_cols, 
-        int num_left_cols,
-        int left_join_cols[],
-        gdf_column **right_cols,
-        int num_right_cols,
-        int right_join_cols[],
-        int num_cols_to_join,
-        int result_num_cols,
-        gdf_column ** result_cols,
+cudf::table construct_join_output_df(
+        cudf::table const& left, 
+        cudf::table const& right,
+        std::vector<std::pair<gdf_size_type, gdf_size_type>> const& columns_in_common,
         gdf_column * left_indices,
         gdf_column * right_indices) {
 
-
-  PUSH_RANGE("LIBGDF_JOIN_OUTPUT", JOIN_COLOR);
+    PUSH_RANGE("LIBGDF_JOIN_OUTPUT", JOIN_COLOR);
     //create left and right input table with columns not joined on
-    std::vector<gdf_column*> lnonjoincol;
-    std::vector<gdf_column*> rnonjoincol;
-    std::set<int> l_join_indices, r_join_indices;
-    for (int i = 0; i < num_cols_to_join; ++i) {
-        l_join_indices.insert(left_join_cols[i]);
-        r_join_indices.insert(right_join_cols[i]);
-    }
-    for (int i = 0; i < num_left_cols; ++i) {
-        if (l_join_indices.find(i) == l_join_indices.end()) {
-            lnonjoincol.push_back(left_cols[i]);
-        }
-    }
-    for (int i = 0; i < num_right_cols; ++i) {
-        if (r_join_indices.find(i) == r_join_indices.end()) {
-            rnonjoincol.push_back(right_cols[i]);
-        }
+    std::vector<gdf_size_type> left_column_indices(left.num_columns());
+    std::vector<gdf_size_type> right_column_indices(right.num_columns());
+    std::vector<gdf_size_type> left_columns_in_common (columns_in_common.size());
+    std::vector<gdf_size_type> right_columns_in_common (columns_in_common.size());
+    std::vector<gdf_size_type> left_nonjoin_indices (left.num_columns() - left_columns_in_common.size());
```

Comment:
Referring to these as "nonjoin_indices" doesn't seem to match with how they are being used. Instead it seems to be the indices of all non-common columns.